src/main/java/com/googlecode/objectify/impl/TypeUtils.java

```diff
@@ -113,11 +125,27 @@ public static boolean isAssignableFrom(Class<?> to, Class<?> from) {
 		return null;
 	}
 
+	@Value
+	private static final class DeclaredAnnotationCacheKey {
+		private final Class<?> onClass;
+		private final Class<? extends Annotation> annotationType;
+	}
+	
+	private static final Map<DeclaredAnnotationCacheKey, Optional<? extends Annotation>> declaredAnnotationCache = Maps.newConcurrentMap();
+	
 	/**
 	 * Get the declared annotation, ignoring any inherited annotations
 	 */
 	public static <A extends Annotation> A getDeclaredAnnotation(Class<?> onClass, Class<A> annotationType) {
-		return getAnnotation(onClass.getDeclaredAnnotations(), annotationType);
+		final DeclaredAnnotationCacheKey key = new DeclaredAnnotationCacheKey(onClass, annotationType);
+		
+		@SuppressWarnings("unchecked")
+		Optional<A> value = (Optional<A>) declaredAnnotationCache.get(key);
```

Comment:
The map is thread safe, but in theory you can compute the cache value more than once. `Map.computeIfAbsent()` would solve this.

Reply:
Java 7. No such method here.